Implement serverless function: invite to jdsb-wrecking-ball.slack.com

kimlisa opened the pull request:
### Purpose:
Serverless function provides link to automated invite to `jdsb-wrecking-ball` workspace

### Slack Slash Command:
`/struggle dev invite`

### Test URL Endpoint:
[https://lisas-first-serverless-func-test.netlify.com/.netlify/functions/inviteme
](https://lisas-first-serverless-func-test.netlify.com/.netlify/functions/inviteme
)
### Tests:
#### slack response:
![Screen Shot 2019-08-12 at 12 23 29 AM](https://user-images.githubusercontent.com/43280172/62850358-7f19ed80-bc97-11e9-9b51-8c4d5f89a2f3.png)

#### browser response:
![Screen Shot 2019-08-12 at 12 30 38 AM](https://user-images.githubusercontent.com/43280172/62850685-770e7d80-bc98-11e9-9f65-6aff07ba2291.png)

kimlisa committed:
Lisa's init

- Create a invitme directory and a main.go
- Changed path and vars for staging to my forked git and my deployed netlify site

kimlisa committed:
Remove .idea files and update gitignore

kimlisa committed:
Test aws-lambda-go example

kimlisa committed:
Testing another aws-lambda-go example

kimlisa committed:
Test aws-lambda-go netlify's example

kimlisa committed:
Test with logs

kimlisa committed:
Test with logs for APIGatewayProxyRequest object

kimlisa committed:
Implement feature: provide link to automated invite to wrecking-ball

- Parse request body
- Create structs for slack message payload metadata
- Register function into slack-app-registry.json

kimlisa commented on inviteme/main.go:
```diff
@@ -0,0 +1,104 @@
+func handler(request events.APIGatewayProxyRequest) (*events.APIGatewayProxyResponse, error) {
+	parsedBody, err := url.ParseQuery(request.Body)
+	if err != nil {
+		return &events.APIGatewayProxyResponse{
```
I just caught that I am not using my actual error messages.

When encountering errors, I should probably be returning a nil Response, and an error that looks something like the following?:

```
errMsg := fmt.Sprintf("[ERROR] %v", err)
return nil, errors.New(errMsg)
```

bhurstGH commented on netlify.toml:
```diff
@@ -3,8 +3,8 @@
-  ROUTER_REGISTRY_URL_ENV_VAR = "https://struggle-slack-app-beta.netlify.com/slack-app-registry.json"
-  GO_IMPORT_PATH = "github.com/junior-dev-struggle-bus/struggle-slack-app"
+  ROUTER_REGISTRY_URL_ENV_VAR = "https://lisas-first-serverless-func-test.netlify.com/slack-app-registry.json"
+  GO_IMPORT_PATH = "github.com/kimlisa/struggle-slack-app"
```
Mark can correct me if I'm wrong, but these should probably be reverted back to the struggle-slack-app-beta URL and JDSB github before merging. :) I don't think it would hurt much right now, but the next time someone else tried to commit changes, they would be pointing to your registry and github.

phoenixcoder replied:
Yes, you'll want to revert this back to the way it was. In the future, you can create a branch called, "lisasdev", and add an entry in the netlify.toml like so:

[context.lisasdev.environment]
ROUTER_REGISTRY_URL_ENV_VAR = "..."
GO_IMPORT_PATH = "..."

phoenixcoder commented on .netlify/state.json:
```diff
@@ -0,0 +1,3 @@
+{
```
Please remove this file. It's netlify generated.

phoenixcoder commented on inviteme/deploy.sh:
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
```
Please remove this file. This looks like a helper file for this specific project?

phoenixcoder commented on public/slack-app-registry.json:
```diff
@@ -13,11 +13,16 @@
+                "description" : "Mark has no idea what this function does.",
+                "manual" : "Hope Brett doesn't kill us."
+            },
+            "inviteme" : {
```
This has to be the name of the function. In this case, it would be `/struggle inviteme`

phoenixcoder commented on public/slack-app-registry.json:
```diff
@@ -13,11 +13,16 @@
+                "manual" : "Hope Brett doesn't kill us."
+            },
+            "inviteme" : {
+                "usage" : "/struggle dev invite",
```
Sorry, I recommended changing it to `/struggle dev invite`, but this would require making a `dev` named function first. So, just keep it to `inviteme` for now, and change this to, `/struggle inviteme`.

phoenixcoder commented on public/slack-app-registry.json:
```diff
@@ -13,11 +13,16 @@
+            "inviteme" : {
+                "usage" : "/struggle dev invite",
+                "description" : "Returns a link that will send automated invite to jdsb-wrecking-ball workspace",
+                "manual" : "Just type `/struggle dev invite` in slack channel"
```
Go ahead and add an "author" entry. You can put your name or alias in there e.g.

"author" : "lisakeem"

phoenixcoder commented on inviteme/main.go:
```diff
@@ -0,0 +1,104 @@
+	imageText     string = "I came in like a wrecking ball"
+	imageAltText  string = "Picture of a cat wrecking neatly stacked books"
+	linkMsg       string = "Click on the link below to get an invite to *jdsb-wrecking-ball* workspace: "
+	myJdsbSlackID        = "UGAA0BJ4R"
```
It doesn't look like this converted to your username in your test unless you blocked it out on purpose?

kimlisa replied:
This is the ID for the jdsb channel, so it won't recognize unless you run command in that channel. I did test to see if it worked using my ID in my personal workspace. Should I change it to wrecking-ball ID?

phoenixcoder replied:
You can keep it this way for now. I'll add a way to change environments through the registry.

phoenixcoder replied:
Added an issue for this: #34

phoenixcoder commented on inviteme/main.go:
```diff
@@ -0,0 +1,104 @@
+	imageLink     string = "http://giphygifs.s3.amazonaws.com/media/5q7XU3MurT0pW/giphy.gif"
+	imageText     string = "I came in like a wrecking ball"
+	imageAltText  string = "Picture of a cat wrecking neatly stacked books"
+	linkMsg       string = "Click on the link below to get an invite to *jdsb-wrecking-ball* workspace: "
```
The type declarations here are redundant. The compiler already knows the type to initialize to based on the type of the value.

phoenixcoder reviewed (changes requested):
Made a first pass and indicated what files and configuration needs to be changed. I did not give a review on the main logic yet. Let's clean up the other things first.

So far, it's looking good!

kimlisa committed:
Make change suggested by code review

- Deleted non production files
- Edited inviteme register usage and included author
- Removed redundant type declarations in inviteme/main.go
- Reverted netlify.toml to point back to original settings

phoenixcoder commented on inviteme/main.go:
```diff
@@ -0,0 +1,104 @@
+	parsedBody, err := url.ParseQuery(request.Body)
+	if err != nil {
+		return &events.APIGatewayProxyResponse{
+			StatusCode: http.StatusBadRequest,
```
Bad request is a 400 error, which blames the user for doing something wrong. This is inaccurate in the context of a Slack app. If no body comes in, that means Slack didn't send the body over. It's the service that screwed up in this case. So, an internal error (http status code 500) is accurate.

However, because of the way Slack reports any status code aside from 200 OK e.g. 400_user_error (no error message), you may want to return 200, then put the error message you want in the body. This would be more friendly to users.

phoenixcoder commented on inviteme/main.go:
```diff
@@ -0,0 +1,104 @@
+)
+
+func handler(request events.APIGatewayProxyRequest) (*events.APIGatewayProxyResponse, error) {
+	parsedBody, err := url.ParseQuery(request.Body)
```
You'll want to authenticate the request since we have no way of controlling function permissions via Netlify. I've provided a simple API to do that, you just need to add a line in your code like so: https://github.com/junior-dev-struggle-bus/struggle-slack-app/blob/895d67adbc49655ac422434b079a904a57cfb96c/umbrella/router.go#L192

You'll need to import the following library:
https://github.com/phoenixcoder/slack-golang-sdk/blob/master/auth/auth.go

To test in your dev environment, you'll need to add your Slack App's Signing Secret to your Netlify environment: https://github.com/phoenixcoder/slack-golang-sdk/blob/de18e751112634d6dfb92c6be24536f1781427ba/auth/auth.go#L21

**Reference:**
Getting your Slack App's Signing Secret: https://api.slack.com/docs/verifying-requests-from-slack#signing_secrets_admin_page
Add environment var to your netlify: https://www.netlify.com/docs/continuous-deployment/#environment-variables

phoenixcoder reviewed (changes requested):
Authenticate the request. Other logic looks good so far.

kimlisa committed:
Make change based on review

- authenticate lambda request
- for errors, returned 200 with desired error messages

phoenixcoder commented on inviteme/slackapi.go:
```diff
@@ -0,0 +1,81 @@
+package main
```
I would not publish these just yet for this PR. When you make something public, you're making a contract with the user and community that you will support this forever and not break them. 

If you want to publish them, please consider moving these to this package first: https://github.com/phoenixcoder/slack-golang-sdk

Otherwise, make everything lowercase for now.

phoenixcoder commented on inviteme/slackapi.go:
```diff
@@ -0,0 +1,81 @@
+	ErrParseRequestBody    = errors.New("unable to parse APIGatewayProxyRequest.Body")
+	ErrDecodingParsedQuery = errors.New("unable to decode parsed query into a struct")
+	ErrJSONMarshal         = errors.New("unable to JSON encode for response")
+	ErrAuthLambdaReq       = errors.New("unable to authenticate lambda request")
```
These errors get shown to the user. You'll probably want to make them more friendly e.g. Oh no...the service screwed up. Sorry about that! Please try it again.

I don't know that any one will understand what an APIGatewayProxyRequest.Body is at first glance.

phoenixcoder commented on inviteme/main.go:
```diff
@@ -0,0 +1,123 @@
+
+// Bad request is a 400 error, which blames the user for doing something wrong
+// Internal error is a 500 error, the service screwed up
+// ie: If no request.Body comes in, that meant Slack didn't send the body over (service error)
```
If we want thorough documentation, you'll need to itemize the reasoning for each error return path. I don't think it's necessary, but if you want to do it for your own edification.

kimlisa replied:
oops hehe, i am used to putting notes all over my code. i shall remove it!

phoenixcoder reviewed (changes requested):
Almost there!

kimlisa committed:
Unexport slackapi.go and use user friendly error messages

phoenixcoder reviewed (approved):
Looks good to me! (LGTM)
